### ISSUE TITLE
Missed optimization: when the value of another variable can be used directly

### ISSUE THREAD

**Issue by ZY546** [llvm/llvm-project#64004]
Test code: https://godbolt.org/z/e97jKPr7e
```C++
int var22=0;
int var23=0;
void test(int var7, int var9, int var11, int var12) {
    var22 = var12 + var9 + var11;
    var23 = var12 + (var9 + var7 + var11);
}
```
IR at -O3
```
  %add = add i32 %var11, %var9
  %add1 = add i32 %add, %var12
  store i32 %add1, ptr @var22, align 4
  %add3 = add i32 %add, %var7
  %add4 = add i32 %add3, %var12
  store i32 %add4, ptr @var23, align 4
  ret void
```
In fact, `var23` can be directly optimized to `var22+var7`, but LLVM does not do this in the IR generated at the O3 optimization level.

**Comment by XChy** [comment]
The transformation seems to be correct.
[alivetv](https://alive2.llvm.org/ce/z/RLpTQJ)

**Comment by XChy** [comment]
Actually, `opt -passes=reassociate,gvn` folds the expression as you [expected](https://godbolt.org/z/qnf9e8vW8). And clang emits the correctly optimized x86 [assembly](https://godbolt.org/z/9a1rT7e5h) with O1 flag, too. I'm not sure why clang emits non-optimized llvmir, but it's clang that is responsible for this problem.

**Comment by ZY546** [comment]
> Actually, `opt -passes=reassociate,gvn` folds the expression as you [expected](https://godbolt.org/z/qnf9e8vW8). And clang emits the correctly optimized x86 [assembly](https://godbolt.org/z/9a1rT7e5h) with O1 flag, too. I'm not sure why clang emits non-optimized llvmir, but it's clang that is responsible for this problem.

With that said, maybe the problem is similar to #63997. But the difference is that this test case is not optimized well even for O3. In fact, at the end of the llvmir optimization, `reassociate,gvn` has been executed, but it turns out that the one time execution does not get the best optimization.
If we want better optimization, perhaps we can modify the `pass manager` to adjust the execution of passes without improving the capabilities of passes?

**Comment by XChy** [comment]
opt -O2 also produces expectedly optimized IR. And `opt` with O2,O3 flag has nearly the same pass [pipeline](https://stackoverflow.com/questions/15548023/clang-optimization-levels) with one in `clang`. 
I don't think one time execution of `reassociate,gvn` cannot get it folded, since `opt -passes=reassociate,gvn` does fold. Maybe the cause is the difference of pass pipeline between opt and clang.

**Comment by ZY546** [comment]
> opt -O2 also produces expectedly optimized IR. And `opt` with O2,O3 flag has nearly the same pass [pipeline](https://stackoverflow.com/questions/15548023/clang-optimization-levels) with one in `clang`. I don't think one time execution of `reassociate,gvn` cannot get it folded, since `opt -passes=reassociate,gvn` does fold. Maybe the cause is the difference of pass pipeline between opt and clang.

https://godbolt.org/z/qbxMeob67  Maybe this result explains my guess better?

**Comment by XChy** [comment]
I see. 
You mean that clang has only one layer of `reassociate, gvn`. Therefore clang just fold one common expression. But 2 layers of optimization could fold it completely, such as `-passes=reassociate,gvn,reassociate,gvn`?
If not refactoring Reassociate and GVN, the easiest solution to this is to add one more `reassociate,gvn` to O2 pipeline.
This is a question whether it's worthwhile. Others' opinions are needed here. @nikic, what's your opinion?

**Comment by ZY546** [comment]
I ran an experiment to see if these changes made any sense:

I randomly generated a [test case](https://godbolt.org/z/v788shEzo) and optimized it with O3 levels, resulting in 181 lines of [llvmir](https://godbolt.org/z/v788shEzo) (only those related to the `test function`, 22-203 lines) and 273 lines of [assembler code](https://godbolt.org/z/v788shEzo).

If we use the following command to optimize the llvmir generated above with O3 level again, we get a further optimized llvmir ([test33.ll](https://godbolt.org/z/MxbxqsE1b)).

`clang -O3 -S -emit-llvm -fno-discard-value-names ./test3.ll -o ./test33.ll`

The newly generated llvmir([test33.ll](https://godbolt.org/z/MxbxqsE1b)) will be reduced to 156 lines (only those related to the `test function` are counted, from 28 to 184 lines).

For further optimized llvmir([test33.ll](https://godbolt.org/z/MxbxqsE1b)), the final assembler code obtained with clang-O3 is 222 lines.

In summary, for this randomly generated test case, it seems that we can use this method to reduce the final generated llvmir from 181 lines to 156 lines and the final generated assembler code from 273 lines to 222 lines.

**Comment by XChy** [comment]
Can I understand that as optimizing the same IR **twice** with O3 flag? If so, that's not a good choice. There's always a balance between compilation time and running time. Though it actually produced better IR, the time consumed is unacceptable for many and some duplicate optimizations make no sense. Sure, you can do so yourself to get the high-quality IR/executable you want.

I think the core problems here are:
- Can we refactor GVN and Reassociate to make them work better ?
- Or is it worthwhile to add multiple `gvn,reassociate` to pipeline (Actually, both of them cost much) ?

**Comment by nikic** [comment]
The NaryReassociate pass probably does what you want.

Running GVN twice is not possible, it is too expensive.

**Comment by ZY546** [comment]
> The NaryReassociate pass probably does what you want.
> 
> Running GVN twice is not possible, it is too expensive.

https://godbolt.org/z/63TGfxcMn  The NaryReassociate pass doesn't seem to work.
The cost of optimization is indeed an issue.

**Comment by XChy** [comment]
GVN is always need to eliminate reduntancy. And single nary-reassociate may be not enough. It does not handle things below currently:
```
foo((a + b) + c);
foo(a + ((b + d) + c));
```
Nary-reassociate check ternary additions at a time, only folding things like:
```
foo((a + b) + c);
foo(d + ((b + c) + a));
```
It's better to add reassociate. [Example here](https://godbolt.org/z/4hY375aeG).

**Comment by ZY546** [comment]
> GVN is always needed to eliminate reduntancy. And single nary-reassociate may be not enough. It does not handle things below currently:
> 
> ```
> foo((a + b) + c);
> foo(a + ((b + d) + c));
> ```
> 
> Nary-reassociate check ternary additions at a time, only folding things like:
> 
> ```
> foo((a + b) + c);
> foo(d + ((b + c) + a));
> ```
> 
> It's better to add reassociate. [Example here](https://godbolt.org/z/4hY375aeG).

It seems a little restrictive. It seems that we can only self-select the combination of multiple Reassocaite and Nary-reassociate for optimization depending on the specific test case.

**Comment by XChy** [comment]
The combination `reassociate, nary-reassociate` covers most cases. And the most of optimizations of LLVM are heuristic, impossible to cover all cases with little cost. No silver bullet here.
Thanks for your issue, maybe it shoud be closed.

**Comment by ZY546** [comment]
I was shocked by a new discovery:
For this test case, if we just change the order of the function arguments, we get the expected optimization:
https://godbolt.org/z/433K3Pcjf
```C++
int var22=0;
int var23=0;
void test(int var12, int var9, int var11, int var7) {
    var22 = var12 + var9 + var11;
    var23 = var12 + (var9 + var7 + var11);
}
```
```
define dso_local void @test(int, int, int, int)(i32 noundef %var12, i32 noundef %var9, i32 noundef %var11, i32 noundef %var7) local_unnamed_addr {
entry:
  %add = add i32 %var9, %var12
  %add1 = add i32 %add, %var11
  store i32 %add1, ptr @var22, align 4
  %add4 = add i32 %add1, %var7
  store i32 %add4, ptr @var23, align 4
  ret void
}
```
While before the execution of `reassociate`, these two test cases have the same llvmir, but the modified test cases are better optimized due to the parameter order.

This is a little beyond my knowledge. Can you help me? @XChy  
Is the optimization result of `reassociate` affected by the order of the parameters? Perhaps it is possible to eliminate this effect to get better optimization?

**Comment by XChy** [comment]
**Reassociate Pass** provides a "rank" for each Value, builds a rank map and canonicalizes operands of instructions by rank. 

When building rank map, it determines the rank of constant/global to be 0, the rank of the **ith** function argument to be **i + 1**, and then distributes incremental rank for other Value by **ReversePostOrderTraversal**.
And by these ranks, **Reassociate Pass** sorts the operands in commutative/reassociable operators.
There are other things like LinearizeExprTree, details of sorting in source code.

Here, you can build rank map:
```
var12 : 2, var9: 3, var11: 4, var7: 5
```
So `var12 + var9 + var11` will be `(var9 + var12) + var11`.
`var12 + (var9 + var7 + var11)` will be `((var9 + var12) + var11) + var7`.
That can be eliminated by GVN directly.

**Comment by ZY546** [comment]
> **Reassociate Pass** provides a "rank" for each Value, builds a rank map and canonicalizes operands of instructions by rank.
> 
> When building rank map, it determines the rank of constant/global to be 0, the rank of the **ith** function argument to be **i + 1**, and then distributes incremental rank for other Value by **ReversePostOrderTraversal**. And by these ranks, **Reassociate Pass** sorts the operands in commutative/reassociable operators. There are other things like LinearizeExprTree, details of sorting in source code.
> 
> Here, you can build rank map:
> 
> ```
> var12 : 2, var9: 3, var11: 4, var7: 5
> ```
> 
> So `var12 + var9 + var11` will be `(var9 + var12) + var11`. `var12 + (var9 + var7 + var11)` will be `((var9 + var12) + var11) + var7`. That can be eliminated by GVN directly.

I see. Thank you very much!
